include/kernels/blas_core.cuh

```diff
@@ -35,15 +36,19 @@ namespace quda
       const int nParity;
       BlasArg(ColorSpinorField &x, ColorSpinorField &y, ColorSpinorField &z, ColorSpinorField &w,
               ColorSpinorField &v, Functor f, int length, int nParity) :
-        kernel_param(dim3(length, nParity, 1)),
+        kernel_param<Functor::use_kernel_arg>(dim3(length, nParity, 1)),
         X(x),
         Y(y),
         Z(z),
         W(w),
         V(v),
         f(f),
         nParity(nParity)
-      { ; }
+      {
+        if constexpr (std::is_same_v<Functor, caxpyxmazMR_<real>>) {
```

Comment:
Is this `static_assert` and the similar one in the reduce_core for the heavy quark residuals necessary?  Can't this `static_assert` be done generically in the kernel launch in `targets/***/tunable_kernel.h`?

Reply:
In `tunable_kernel.h`, we only have access to the `Arg`, and we don't have all the functor names declared.  It's possible to move the `static_assert` there, if we add `using MyFunctor = Functor;` in `BlasArg`/`ReduceArg`, and declare the three functors in `tunable_kernel.h`.

I prefer putting the `static_assert` here, because it makes the intention clear, with the implementation of the functor and its constraint in a single file.  At the very least, it complements the less strict `static_assert` in `Blas_`/`Reduce_`, which by itself does not ensure code correctness.

On the other hand, we can merge this `static_assert` here with the one in `Blas_`/`Reduce_` below, if we have access to the functor from `BlasArg` by way of `using MyFunctor = Functor`.  What do you think of this?